doc: active record and repository patterns (#6923)
quarkusio/quarkus: docs/src/main/asciidoc/hibernate-orm-panache.adoc

NOTE: the `list()` method might be surprising at first. It takes fragments of HQL (JP-QL) queries and contextualizes the rest. That makes for very concise but yet readable code.

> NOTE: what you just see is the link:https://www.martinfowler.com/eaaCatalog/activeRecord.html[active record pattern], sometimes just called the entity pattern.

Reviewer: ```suggestion
NOTE: what was described above is a essentially the link:https://www.martinfowler.com/eaaCatalog/activeRecord.html[active record pattern], sometimes just called the entity pattern.
```

Author: `is a essentially the` isn't the `a` needed ?

Reviewer: the `a` was indeed a mistake :)

> Hibernate with Panache also provides the more classical link:https://martinfowler.com/eaaCatalog/repository.html[repository pattern] via `PanacheRepository`.

Reviewer: ```suggestion
Hibernate with Panache also for the use of the more classical link:https://martinfowler.com/eaaCatalog/repository.html[repository pattern] via `PanacheRepository`.
```

Author: This suggestion look like a verb is missing ... Di you means `allow for the use` ?

Reviewer: Yup :)

NOTE: The `stream` methods require a transaction to work.

=== Adding entity methods

> In general, we recommend not adding custom queries for your entities outside of the entities themselves,

Reviewer 2: style wise, try and keep your sentences light. By definition what you write is what you recommend.
Also "in general", "usually" etc bring no value.
Also be positive, negations are always triggering some negative loops in the brain
Keep sentences short.

Add custom queries on your entities outside of the entities themselves. 
That way, you and your co-workers can find the easily and queries are right next to the object they operate on.
Adding them as static methods in your entity class is the Panache Active Record way.

in your entity class:

[source,java]
----
@Entity
public class Person extends PanacheEntity {
    public String name;
    public LocalDate birth;
    public Status status;

    public static Person findByName(String name){
        return find("name", name).firstResult();
    }

    public static List<Person> findAlive(){
        return list("status", Status.Alive);
    }

    public static void deleteStefs(){
        delete("name", "Stef");
    }
}
----

== Solution 2: using the repository pattern

=== Defining your repository

When using Repositories, you can get the exact same convenient methods as with the active record pattern injected in your Repository,
by making them implements `PanacheRepository`:

[source,java]
----
@ApplicationScoped
public class PersonRepository implements PanacheRepository<Person> {

   // put your custom logic here as instance methods

   public Person findByName(String name){
       return find("name", name).firstResult();
   }

   public List<Person> findAlive(){
       return list("status", Status.Alive);
   }

   public void deleteStefs(){
       delete("name", "Stef");
  }
}
----

> Absolutely all the operations that are defined on `PanacheEntityBase` are available on your repository, so using it

Reviewer 2: s/Absolutely a/A/

Just be factual, not theatrical.

is exactly the same except you need to inject it:

[source,java]
----
@Inject
PersonRepository personRepository;

@GET
public long count(){
    return personRepository.count();
}
----

> Even with repositories, you can keep your entities as subclasses of `PanacheEntity` in order to get the ID and public fields working,

Reviewer 2: Split sentence in 2, it's too long and hard to read.

NOTE: The `stream` methods require a transaction to work.

=== Adding entity methods

Add custom queries on your entities inside the entities themselves.

> That way, you and your co-workers can find them easily, and queries are right next to the object they operate on.

Reviewer: ```suggestion
That way, you and your co-workers can find them easily, and queries are co-located with the object they operate on.
```

    }

    public static List<Person> findAlive(){
        return list("status", Status.Alive);
    }

    public static void deleteStefs(){
        delete("name", "Stef");
    }
}
----

== Solution 2: using the repository pattern

=== Defining your entity

When using the repository pattern, you can define your entities as regular JPA entities.

[source,java]
----
@Entity
public class Person {
    @Id @GeneratedValue private Long id;
    private String name;
    private LocalDate birth;
    private Status status;

    public Long getId(){
        return id;
    }
    public void setId(Long id){
        this.id = id;
    }
    public String getName() {
        return name;
    }
    public void setName(String name) {
        this.name = name;
    }
    public LocalDate getBirth() {
        return birth;
    }
    public void setBirth(LocalDate birth) {
        this.birth = birth;
    }
    public Status getStatus() {
        return status;
    }
    public void setStatus(Status status) {
        this.status = status;
    }
}
----

TIP: If you don't want to bother defining getters/setters for your entities, you can make them extend `PanacheEntityBase` and

> Quarkus will generates them for you. You can even extend `PanacheEntity` and take advantage of the default ID it provides.

Reviewer: ```suggestion
Quarkus will generate them for you. You can even extend `PanacheEntity` and take advantage of the default ID it provides.
```

=== Defining your repository

> When using Repositories, you can get the exact same convenient methods as with the active record pattern, injected in your Repository,

Reviewer: ```suggestion
When using Repositories, you get the exact same convenient methods as with the active record pattern, injected in your Repository,
```

by making them implements `PanacheRepository`:

[source,java]
----
@ApplicationScoped
public class PersonRepository implements PanacheRepository<Person> {

   // put your custom logic here as instance methods

   public Person findByName(String name){
       return find("name", name).firstResult();
   }

   public List<Person> findAlive(){
       return list("status", Status.Alive);
   }

   public void deleteStefs(){
       delete("name", "Stef");
  }
}
----

All the operations that are defined on `PanacheEntityBase` are available on your repository, so using it
is exactly the same as using the active record pattern, except you need to inject it:

[source,java]
----
@Inject
PersonRepository personRepository;

@GET
public long count(){
    return personRepository.count();
}
----

=== Most useful operations

> Once you have written your repository, here are the most common operations you will be able to do:

Reviewer: ```suggestion
Once you have written your repository, here are the most common operations you will be able to perform:
```

person = personRepository.findById(personId);

// finding a specific person by ID via an Optional
Optional<Person> optional = personRepository.findByIdOptional(personId);
person = optional.orElseThrow(() -> new NotFoundException());

// finding all living persons
List<Person> livingPersons = personRepository.list("status", Status.Alive);

// counting all persons
long countAll = personRepository.count();

// counting all living persons
long countAlive = personRepository.count("status", Status.Alive);

// delete all living persons
personRepository.delete("status", Status.Alive);

// delete all persons
personRepository.deleteAll();

// update all living persons
personRepository.update("name = 'Moral' where status = ?1", Status.Alive);

----

All `list` methods have equivalent `stream` versions.

[source,java]
----
Stream<Person> persons = personRepository.streamAll();
List<String> namesButEmmanuels = persons
    .map(p -> p.name.toLowerCase() )
    .filter( n -> ! "emmanuel".equals(n) )
    .collect(Collectors.toList());
----

NOTE: The `stream` methods require a transaction to work.

NOTE: The rest of the documentation show usages based on the active record pattern only,

> but they can be done with the repository pattern as well.

Reviewer: ```suggestion
but keep in mind that can be performed with the repository pattern as well.
```

Author: Isn't a they missing ?
`but keep in mind that can be performed ` => `but keep in mind that they can be performed ` ?

Reviewer: Yeah... And now you know why I don't usually review docs :). My hands are always a step behind my thought  :(

Author: ;)

> We just don't feel the need to duplicate all the documentation for both patterns.

Reviewer: ```suggestion
The repository pattern examples have been omitted for brevity. 
```
